src/ripple/rpc/impl/TransactionSign.cpp

```diff
@@ -826,40 +905,37 @@ Json::Value transactionSignFor (
             return rpcError (err);
     }
 
-    // Make sure the STTx makes a legitimate Transaction.
-    std::pair <Json::Value, Transaction::pointer> txn =
-        transactionConstructImpl (preprocResult.second,
-            ledger->rules(), app);
-
-    if (!txn.second)
-        return txn.first;
-
-    Json::Value json = transactionFormatResultImpl (txn.second);
-    if (RPC::contains_error (json))
-        return json;
-
-    // Finally, do what we were called for: return a Signers array.  Build
-    // a Signer object to insert into the Signers array.
-    Json::Value signer (Json::objectValue);
-
-    signer[sfAccount.getJsonName ()] = toBase58 (*signerAccountID);
+    // Inject the newly generated signature into tx_json.Signers.
+    auto& sttx = preprocResult.second;
+    {
+        // Make the signer object that we'll inject.
+        STObject signer (sfSigner);
+        signer[sfAccount] = *signerAccountID;
+        signer.setFieldVL (sfTxnSignature, multiSignature);
+        signer.setFieldVL (sfSigningPubKey, multiSignPubKey.getAccountPublic());
 
-    signer[sfSigningPubKey.getJsonName ()] =
-        strHex (multiSignPubKey.getAccountPublic ());
+        // If there is not yet a Signers array, make one.
+        if (!sttx->isFieldPresent (sfSigners))
+            sttx->setFieldArray (sfSigners, {});
 
-    signer[sfTxnSignature.getJsonName ()] = strHex (multiSignature);
+        auto& signers = sttx->peekFieldArray (sfSigners);
+        signers.emplace_back (std::move (signer));
 
-    // Give the Signer an object name and put it in the Signers array.
-    Json::Value nameSigner (Json::objectValue);
-    nameSigner[sfSigner.getJsonName ()] = std::move (signer);
+        // The array must be sorted and validated.
+        auto err = sortAndValidateSigners (signers, (*sttx)[sfAccount]);
+        if (RPC::contains_error (err))
+            return std::move (err);
```

Comment:
@scottschurr If you `return err;` instead RVO kicks in. The `std::move` prevents the compiler from using RVO.